Blog: Connecting to a service without SBO (#5915)
redhat-developer/odo: docs/website/blog/2022-06-30-binding-database-service-without-sbo.md

Or for sharding setup:
  $ kubectl run -i --rm --tty percona-client --image=percona/percona-server-mongodb:4.4 --restart=Never -- mongo "mongodb://${ADMIN_USER}:${ADMIN_PASSWORD}@minimal-cluster-mongos.restapi-mongodb.svc.cluster.local/admin?ssl=false"
```

Wait for the pods to come up, this might take a few minutes:
```sh
$ kubectl get pods
NAME                                    READY   STATUS    RESTARTS   AGE
minimal-cluster-cfg-0                   1/1     Running   0          3m6s
minimal-cluster-mongos-0                1/1     Running   0          3m5s
minimal-cluster-rs0-0                   1/1     Running   0          3m5s
my-op-psmdb-operator-69d88f479c-7hj5m   1/1     Running   0          3m34s
```

## Download the devfile.yaml
6. Run `odo init` to fetch the necessary devfile.
```sh
odo init --devfile go --name places
```

7. Edit the 'runtime' component container in devfile to add information such as username, password, and host required to connect to the mongodb service.
```yaml
components:
- container:
    ...
    ...
    env:
    - name: username
      value: userAdmin
    - name: password

>       value: userAdmin123456

Author: Does devfile support dynamically obtaining values? I mean if we can put `kubectl get secrets minimal-secret -ojsonpath='{.data.username}' | base64 -d` inside the devfile, we can avoid having to hard code password.

>       value: userAdmin123456

Author: I could add a note acknowledging that it's a bad practice, but this is for example purpose only, or I could find a way to fetch password from the code itself to avoid this practice altogether.

>       value: userAdmin123456

Reviewer: Yes, you could start with this solution, by indicating from the very beginning this is not a secure solution, and then continue with more secure solutions (and probaly more difficult ones). That would show the interest of using SBO.

- The simplest solution would be to have an `envFrom` field in the Devfile's container, but this field does not exist. Is it worth asking Devfile team to add it? I'm not sure
- You could pass the secret name and the secret keys as environment variables, and the program would be responsible for getting the values from the secret. That means the program needs access to the Kubernetes API, which opens security breaches

>       value: userAdmin123456

Author: >The simplest solution would be to have an envFrom field in the Devfile's container, but this field does not exist. Is it worth asking Devfile team to add it? I'm not sure

When you say `envFrom`, do you mean devfile can evaluate the value from the environment variable exposed in your local system?
```yaml
env:
- name: password
  value:
    envFrom: $MONGODB_PASSWORD
```
Something like this? Or more like this https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/pods/inject/pod-secret-envFrom.yaml?

>You could pass the secret name and the secret keys as environment variables, and the program would be responsible for getting the values from the secret. That means the program needs access to the Kubernetes API, which opens security breaches

I think asking the program to be responsible for fetching everything would defeat the purpose of this blog. I think I'll go with a note for now, and it's like you said, it might encourage users to try SBO.

>       value: userAdmin123456

Reviewer: ```
> 
> Something like this? Or more like this https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/pods/inject/pod-secret-envFrom.yaml?
> 

Yes, like this one

>       value: userAdmin123456

Author: But then devfile would have to interact with the k8s API and that does not make a lot of sense to me from the design standpoint. I don't think devfile is dependent on the K8s API right now, and I'm not sure if they'd be willing to create a dependency now. Can we take this up in cabal?

---
title: Binding to a database service without the Service Binding Operator
author: Parthvi Vala
author_url: https://github.com/valaparthvi
author_image_url: https://github.com/valaparthvi.png
tags: ["binding"]
slug: binding-database-service-without-sbo
---

How to bind your application to a database service without the Service Binding Operator.

<!--truncate-->

There are a few ways of binding your application to a database service with the help of odo. The recommended way is with the help of Service Binding Operator(SBO), but it is also possible to bind without it, and this blog will show you how.

## Architecture
We have a simple CRUD application built in Go that can create/list/update/delete a place. This application requires connecting to a MongoDB database in order to function correctly, which will be deployed as a microservice on the cluster.

## Prerequisites:
This blog assumes:
- [odo v3.0.0-beta1](https://github.com/redhat-developer/odo/releases/tag/v3.0.0-beta1)

> - you have access to a Kubernetes or OpenShift cluster to be able to install the MongoDB operator.

Reviewer: ```suggestion
- you have access to a Kubernetes or OpenShift cluster to be able to install the MongoDB database.
```

But I'm not sure this sentence makes sense, now you've removed the "admin" access necessity: installing the MongoDB database does not need more rights than deploying the app with `odo`

- you have _Helm_ installed on your system. See https://helm.sh/docs/intro/install/ for installation instructions.

## (Optional) Setting up the namespace
0. We will create a new namespace to deploy our application in, with the help of odo.
```sh
odo create namespace restapi-mongodb
```

## Setting up the MongoDB microservice
We are going to use the Bitnami's helm charts for creating our MongoDB database.

1. Add the Bitnami's Helm charts repository and make your Helm client up to date with it:
```sh
helm repo add bitnami https://charts.bitnami.com/bitnami && helm repo update
```

2. Export the necessary environment variables:
```sh

> export MY_MONGODB_ROOT_USERNAME=root

Reviewer: I'm pretty sure the exports are not necessary.

They would be necessary if you used `odo dev --var MY_MONGODB_ROOT_USERNAME`, but using `odo dev --var PASSWORD=$MY_MONGODB_ROOT_USERNAME `, you just need to declare the variables into the shell session, without exporting them.

Note that exporting a variable (especially a password) into a shell gives access to it from *all* the processes you start from your shell. This can be a security breach, it must be used cautiously.

export MY_MONGODB_ROOT_PASSWORD=my-super-secret-root-password
export MY_MONGODB_USERNAME=my-app-username
export MY_MONGODB_PASSWORD=my-app-super-secret-password
export MY_MONGODB_DATABASE=my-app
```

3. Create the MongoDB service.
```sh
helm install mongodb bitnami/mongodb \
  --set auth.rootPassword=$MY_MONGODB_ROOT_PASSWORD \
  --set auth.username=$MY_MONGODB_USERNAME \
  --set auth.password=$MY_MONGODB_PASSWORD \
  --set auth.database=$MY_MONGODB_DATABASE
```

<details>
<summary>Expected output:</summary>

```sh
$ helm install mongodb bitnami/mongodb \
  --set auth.rootPassword=$MY_MONGODB_ROOT_PASSWORD \
  --set auth.username=$MY_MONGODB_USERNAME \
  --set auth.password=$MY_MONGODB_PASSWORD \
  --set auth.database=$MY_MONGODB_DATABASE
NAME: mongodb
LAST DEPLOYED: Tue Jul  5 15:53:40 2022
NAMESPACE: restapi-mongodb
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
CHART NAME: mongodb
CHART VERSION: 12.1.24
APP VERSION: 5.0.9

** Please be patient while the chart is being deployed **

MongoDB&reg; can be accessed on the following DNS name(s) and ports from within your cluster:

    mongodb.restapi-mongodb.svc.cluster.local

To get the root password run:

>     export MONGODB_ROOT_PASSWORD=$(kubectl get secret --namespace restapi-mongodb mongodb -o jsonpath="{.data.mongodb-root-password}" | base64 -d)

Reviewer: Is it different from MY_MONGODB_ROOT_PASSWORD?

Author: No, it's not.

---

> title: Binding to a database service without SBO

Reviewer 2: How about making the SBO acronym explicit in the page title?

```suggestion
title: Binding to a database service without the Service Binding Operator
```

author: Parthvi Vala
author_url: https://github.com/valaparthvi
author_image_url: https://github.com/valaparthvi.png
tags: ["binding"]
slug: binding-database-service-without-sbo
---

> How to bind to your application to a database service without the Service Binding Operator.

Reviewer 2: ```suggestion
How to bind your application to a database service without the Service Binding Operator.
```

<!--truncate-->

There are a few ways of binding your application to a database service with the help of odo. The recommended way is with the help of Service Binding Operator(SBO), but it is also possible to bind without it, and this blog will show you how.

## Architecture
We have a simple CRUD application built in Go that can create/list/update/delete a place. This application requires connecting to a MongoDB database in order to function correctly, which will be deployed as a microservice on the cluster.

## Prerequisites:
This blog assumes:
- [odo v3.0.0-beta1](https://github.com/redhat-developer/odo/releases/tag/v3.0.0-beta1)

> - you have admin access to a Kubernetes or OpenShift cluster to be able to install the MongoDB operator.

Reviewer 2: Indeed, this should not be necessary.

- you have _Helm_ installed on your system. See https://helm.sh/docs/intro/install/ for installation instructions.

## (Optional) Setting up the namespace
0. We will create a new namespace to deploy our application in, with the help of odo.
```sh
odo create namespace restapi-mongodb
```

## Setting up the application
1. Clone the repository, and cd into it.
```sh
git clone https://github.com/valaparthvi/restapi-mongodb-odo.git && cd restapi-mongodb-odo
```

## Export environment variables

> 2. Export the necessary environment variables:

Author: I agree, but then if we want to maintain the order of `odo init` > `odo dev` > deploy charts > wait for odo dev to detect changes, it becomes necessary to export these variables at this point

> 2. Export the necessary environment variables:

Reviewer: The order 
- odo init
- odo dev
- deploy chart
is just an example provided by Tomas, I'm not sure it must be respected at all price.

By first deploying the chart (including exportint variables), then running odo init and odo dev, the flow seems more understandable to me

> 2. Export the necessary environment variables:

Reviewer 3: > is just an example provided by Tomas, I'm not sure it must be respected at all price.
> 
> 
> By first deploying the chart (including exportint variables), then running odo init and odo dev, the flow seems more understandable to me

I like the current approach of enabling Helm repo first. But I don't agree with the idea of spinning up Helm chart using it before doing `odo init`. If it were me, I would first show that your Go application is deployed using `odo dev`, but `curl` request made to it is throwing 404 or some error. Thereafter, we spin up the MongoDB database required by the application.

Or maybe, do an `odo dev` without all the variables which would help one connect to the MongoDB database. Make a `curl` call that returns 404 or some error. Now add information about connection to MongoDB and see how the application returns expected output.

I'm saying this because that's how I think an application developer's flow would be like. But then I have not a lot of application development myself, so correct me if I'm mistaken.

> 2. Export the necessary environment variables:

Author: > Or maybe, do an `odo dev` without all the variables which would help one connect to the MongoDB database. Make a `curl` call that returns 404 or some error. Now add information about connection to MongoDB and see how the application returns expected output.
> 
> I'm saying this because that's how I think an application developer's flow would be like. But then I have not a lot of application development myself, so correct me if I'm mistaken.

I agree with you, but I can't find a flow that works for everyone. For now, I have added a note stating that `odo dev` will fail if run before adding connection information.

> 2. Export the necessary environment variables:

Reviewer 3: Works for me. I discussed this with @feloy and agree with the flow we have now.

```sh
export MY_MONGODB_ROOT_USERNAME=root
export MY_MONGODB_ROOT_PASSWORD=my-super-secret-root-password
export MY_MONGODB_USERNAME=my-app-username
export MY_MONGODB_PASSWORD=my-app-super-secret-password
export MY_MONGODB_DATABASE=my-app
```

## Download the devfile.yaml
3. Run `odo init` to fetch the necessary devfile.
```sh

> odo init --devfile go --name places

Reviewer 2: There is already a `devfile.yaml` file after cloning your repo. So running `odo init` complains:
```
❯ odo init --devfile go --name places
 ✗  a devfile already exists in the current directory
```

Author: It should be fixed now, I'd accidentally pushed devfile.yaml

```

## Deploy the application
4. Run `odo dev` to deploy the application on the cluster.
```sh

> odo dev --var PASSWORD=$MY_MONGODB_ROOT_PASSWORD --var USERNAME=$MY_MONGODB_ROOT_USERNAME

Reviewer: Same here, I think it is premature to pass these variables, as they are not existing in the Devfile yet and we are not using them. I think it would be better to wait for steps 7.x before to do this

> odo dev --var PASSWORD=$MY_MONGODB_ROOT_PASSWORD --var USERNAME=$MY_MONGODB_ROOT_USERNAME

Reviewer: Same here, I think it is premature to pass these variables, as they are not existing in the Devfile yet and we are not using them. I think it would be better to wait for steps 7.x before to do this

```
</details>

Notice the resources(sevice, deployment, and secrets) that are deployed.

Wait for the pods to come up, this might take a few minutes:
```sh
$ kubectl get pods
NAME                       READY   STATUS    RESTARTS   AGE
mongodb-85fff797f6-fnwvl   1/1     Running   0          63s
```

## Adding the connection information to devfile.yaml
There are 3 changes that we will need to make to our devfile:

7.1 Change the `schemaVersion` of devfile to 2.2.0. This change is required to use certain features of the devfile.
```yaml
schemaVersion: 2.2.0
```
Please note that this change is only necessary because we are using [devfile variable substitution](https://odo.dev/docs/command-reference/dev/#substituting-variables).

7.2 Add a `variables` field in the devfile.
```yaml
variables:
  PASSWORD: password
  USERNAME: user
```
7.3 Edit the 'runtime' container component in devfile to add information such as username, password, and host required to connect to the MongoDB service.
```yaml
components:
- container:
    ...
    ...
    env:
    - name: username
      value: "{{USERNAME}}"
    - name: password
      value: "{{PASSWORD}}"
    - name: host

>       value: "mongodb"

Reviewer 2: The database host could be a variable too, I think. Its default value could still be `mongodb` though.

- you have access to a Kubernetes or OpenShift cluster.
- you have _Helm_ installed on your system. See https://helm.sh/docs/intro/install/ for installation instructions.

## (Optional) Setting up the namespace
0. We will create a new namespace to deploy our application in, with the help of odo.
```sh
odo create namespace restapi-mongodb
```

## Setting up the MongoDB microservice
We are going to use the Bitnami's helm charts for creating our MongoDB database.

1. Add the Bitnami's Helm charts repository and make your Helm client up to date with it:
```sh
helm repo add bitnami https://charts.bitnami.com/bitnami && helm repo update
```

2. Declare the necessary environment variables:
```sh
MY_MONGODB_ROOT_USERNAME=root
MY_MONGODB_ROOT_PASSWORD=my-super-secret-root-password
MY_MONGODB_USERNAME=my-app-username
MY_MONGODB_PASSWORD=my-app-super-secret-password
MY_MONGODB_DATABASE=my-app
```

3. Create the MongoDB service.
```sh
helm install mongodb bitnami/mongodb \
  --set auth.rootPassword=$MY_MONGODB_ROOT_PASSWORD \
  --set auth.username=$MY_MONGODB_USERNAME \
  --set auth.password=$MY_MONGODB_PASSWORD \

>   --set auth.database=$MY_MONGODB_DATABASE

Reviewer: My bad, I used `kubectl create namespace ...` instead, you can ignore most of my comments :(

  USERNAME: user
  HOST: host
```
6.3 Edit the 'runtime' container component in devfile to add information such as username, password, and host required to connect to the MongoDB service.
```yaml
components:
- container:
    ...
    ...
    env:
    - name: username
      value: "{{USERNAME}}"
    - name: password
      value: "{{PASSWORD}}"
    - name: host
      value: "{{HOST}}"
  name: runtime
```

The values for _username_, _password_, and _host_ will be passed to devfile.yaml with the `--var` flag when we run the `odo dev` command.

<details>
<summary>Your final devfile.yaml should look something like this:</summary>

```yaml
commands:
- exec:
    commandLine: GOCACHE=${PROJECT_SOURCE}/.cache go build main.go
    component: runtime
    group:
      isDefault: true
      kind: build
    hotReloadCapable: false
    workingDir: ${PROJECT_SOURCE}
  id: build
- exec:
    commandLine: ./main
    component: runtime
    group:
      isDefault: true
      kind: run
    hotReloadCapable: false
    workingDir: ${PROJECT_SOURCE}
  id: run
components:
- container:
    dedicatedPod: false
    endpoints:
    - name: port-3000-tcp
      protocol: tcp
      secure: false

>       targetPort: 3000

Reviewer: The default port is 8080 in the Devfile I downloaded with `odo init`. It owuld be helpful to indicate to the reader to change it

Reviewer 2: Instead of changing the port in the Devfile, wouldn't it be more straightforward to change the default port in the [application source code](https://github.com/valaparthvi/restapi-mongodb-odo/blob/main/config.yml#L2) to use 8080 as default port instead? This would mean one less manual operation for the readers.

    image: golang:latest
    memoryLimit: 1024Mi
    mountSources: true
    env:
    - name: username
      value: "{{USERNAME}}"
    - name: password
      value: "{{PASSWORD}}"
    - name: host
      value: "{{HOST}}"
  name: runtime
variables:
  PASSWORD: password
  USERNAME: user
  HOST: host
metadata:
  description: Stack with the latest Go version
  displayName: Go Runtime
  icon: https://raw.githubusercontent.com/devfile-samples/devfile-stack-icons/main/golang.svg
  language: go
  name: restapi
  projectType: go
  tags:
  - Go
  version: 1.0.0
schemaVersion: 2.2.0
starterProjects:
- git:
    checkoutFrom:
      revision: main
    remotes:
      origin: https://github.com/devfile-samples/devfile-stack-go.git
  name: go-starter
```
</details>

> ## Deploy the application

Reviewer: You would need to execute this command at some point before:

```
odo set namespace restapi-mongodb
```

7. Run `odo dev` to deploy the application on the cluster.
```sh

> odo dev --var PASSWORD=$MY_MONGODB_ROOT_PASSWORD --var USERNAME=$MY_MONGODB_ROOT_USERNAME  --var HOST="mongodb"

Reviewer 2: Surely nitpicking, but would it be possible to display this command under multiple lines? Otherwise, we need to scroll horizontally to see the full command line:

```suggestion
odo dev \
  --var PASSWORD=$MY_MONGODB_ROOT_PASSWORD \
  --var USERNAME=$MY_MONGODB_ROOT_USERNAME \
  --var HOST="mongodb"
```

 •  Executing the application (command: run)  ...

Your application is now running on the cluster

 - Forwarding from 127.0.0.1:40001 -> 3000

Watching for changes in the current directory /home/pvala/restapi-mongodb-odo
Press Ctrl+c to exit `odo dev` and delete resources from the cluster
```
</details>

## Accessing the application
8. Run the following curl command to test the application:
```sh
curl 127.0.0.1:40001/api/places
```
This will return a _null_ response since the database is currently empty, but it also means that we have successfully connected to our database application.

9. Add some data to the database:
```sh
curl -sSL -XPOST -d '{"title": "Agra", "description": "Land of Tajmahal"}' 127.0.0.1:40001/api/places
```

10. Fetch the list of places again:
```sh
$ curl 127.0.0.1:40001/api/places

{"id":"62c2a0659fa147e382a4db31","title":"Agra","description":"Land of Tajmahal"}
```

### List of available API endpoints
- GET `/api/places` - List all places
- POST `/api/places` - Add a new place
- PUT `/api/places` - Update a place
- GET `/api/places/<id>` - Fetch place with id `<id>`
- DELETE `/api/places/<id>` - Delete place with id `<id>`

## Conclusion

> To conclude this blog, it is possible to connect your application with another microservice without the Service Binding Operator if you have the correct connection information. Using the Service Binding Operator with a [Bindable Operator](https://github.com/redhat-developer/service-binding-operator#known-bindable-operators) makes it easy for you to not care about finding the connection information and ease the binding.

Reviewer: You could add to the conclusion that it is up to the reader to manage the secrets stored in the environment variables MY_MONGODB_ROOT_PASSWORD and MY_MONGODB_ROOT_USERNAME, so they are accessible every time the following command is executed:

```
odo dev --var PASSWORD=$MY_MONGODB_ROOT_PASSWORD --var USERNAME=$MY_MONGODB_ROOT_USERNAME --var HOST="mongodb"
```

Author: I've added this note in the section where we introduce these variables for the first time, I think it makes more sense to put them there than in the conclusion.

Watching for changes in the current directory /home/pvala/restapi-mongodb-odo
Press Ctrl+c to exit `odo dev` and delete resources from the cluster

```
</details>

## Adding the connection information to devfile.yaml
There are a few changes that we will need to make to our devfile:

6.1 Change the `schemaVersion` of devfile to 2.2.0.
```yaml
schemaVersion: 2.2.0
```
Please note that this change is only necessary because we are using [devfile variable substitution](/docs/command-reference/dev#substituting-variables).

6.2 Add a `variables` field in the devfile.
```yaml
variables:
  PASSWORD: password
  USERNAME: user
  HOST: host
```
6.3 Edit the 'runtime' container component in devfile to add information such as username, password, and host required to connect to the MongoDB service.
```yaml
components:
- container:
    ...
    ...
    env:
    - name: username
      value: "{{USERNAME}}"
    - name: password
      value: "{{PASSWORD}}"
    - name: host
      value: "{{HOST}}"
  name: runtime
```

The values for _username_, _password_, and _host_ will be passed to devfile.yaml with the `--var` flag when we run the `odo dev` command.

> 6.4 If you have not already set the 'runtime' container component port to 3000, now is the time to do so.

Reviewer 2: Instead of changing the port in the Devfile, wouldn't it be more straightforward to change the default port in the [application source code](https://github.com/valaparthvi/restapi-mongodb-odo/blob/main/config.yml#L2) to use 8080 as default port instead? This would mean one less manual operation for the readers.

Author: You're right.
